source.go

```diff
@@ -240,30 +246,45 @@ func (a *sourcePluginAdapter) runAck(ctx context.Context, stream cpluginv1.Sourc
 	}
 }
 
-func (a *sourcePluginAdapter) Stop(context.Context, cpluginv1.SourceStopRequest) (cpluginv1.SourceStopResponse, error) {
+func (a *sourcePluginAdapter) Stop(ctx context.Context, _ cpluginv1.SourceStopRequest) (cpluginv1.SourceStopResponse, error) {
 	// stop reading new messages
 	a.openCancel()
 	a.readCancel()
 
-	// TODO timeout for badly written connectors
-	<-a.readDone // wait for read to actually stop running
+	// wait for read to actually stop running with a timeout, in case the
+	// connector gets stuck
+	waitCtx, cancel := context.WithTimeout(ctx, stopTimeout)
+	defer cancel()
+
+	err := a.waitForClose(waitCtx, a.readDone)
+	if err != nil {
+		Logger(ctx).Warn().Err(err).Msg("failed to wait for Read to stop running")
```

Comment:
Nitpick: unless `waitForClose` already does it, it would be good to mention the timeout value in the logs.